File: app/src/main/java/com/example/myapplication/core/progress/WeeklyInsightEngine.kt
package com.example.myapplication.core.progress

import com.example.myapplication.core.feedback.WorkoutDifficulty
import com.example.myapplication.core.feedback.WorkoutFeedback
import com.example.myapplication.core.model.WorkoutHistoryEntry
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.temporal.TemporalAdjusters
import kotlin.math.abs
import kotlin.math.roundToInt

sealed interface WeeklyInsight {
    val messageVi: String

    data class AdherenceTrend(override val messageVi: String) : WeeklyInsight
    data class ReliableWeekday(override val messageVi: String) : WeeklyInsight
    data class DifficultyTrend(override val messageVi: String) : WeeklyInsight
    data class TimeBudgetPattern(override val messageVi: String) : WeeklyInsight
    data class ScheduleDrift(override val messageVi: String) : WeeklyInsight
}

object WeeklyInsightEngine {
    fun generate(
        history: List<WorkoutHistoryEntry>,
        feedback: List<WorkoutFeedback>,
        todayEpochDay: Long,
    ): List<WeeklyInsight> {
        val currentMonday = LocalDate.ofEpochDay(todayEpochDay)
            .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
            .toEpochDay()
        val windowStart = currentMonday - 28
        val evidence = history.filter { it.dueEpochDay in windowStart until currentMonday }
        val byWeek = evidence.groupBy { weekStart(it.dueEpochDay) }.toSortedMap()
        if (byWeek.size < 2) return emptyList()

        return buildList {
            adherenceInsight(byWeek)?.let(::add)
            reliableWeekdayInsight(evidence)?.let(::add)
            difficultyInsight(evidence, feedback)?.let(::add)
            timeBudgetInsight(evidence)?.let(::add)
            scheduleDriftInsight(evidence)?.let(::add)
        }.take(3)
    }

    private fun adherenceInsight(
        byWeek: Map<Long, List<WorkoutHistoryEntry>>,
    ): WeeklyInsight.AdherenceTrend? {
        val rates = byWeek.values.map { week ->
            week.count { it.completedEpochDay != null } * 100.0 / week.size
        }
        val split = rates.size / 2
        if (split == 0) return null
        val older = rates.take(split).average()
        val newer = rates.drop(split).average()
        val difference = (newer - older).roundToInt()
        if (abs(difference) < 15) return null
        val direction = if (difference > 0) "tăng" else "giảm"
        return WeeklyInsight.AdherenceTrend(
            "Tỷ lệ bám lịch $direction ${abs(difference)} điểm phần trăm qua ${rates.size} tuần đầy đủ.",
        )
    }

    private fun reliableWeekdayInsight(
        evidence: List<WorkoutHistoryEntry>,
    ): WeeklyInsight.ReliableWeekday? {
        val candidate = evidence.groupBy { LocalDate.ofEpochDay(it.dueEpochDay).dayOfWeek }
            .filterValues { it.size >= 3 }
            .entries
            .sortedWith(
                compareByDescending<Map.Entry<DayOfWeek, List<WorkoutHistoryEntry>>> {
                    it.value.count { row -> row.completedEpochDay != null }.toDouble() / it.value.size
                }.thenBy { it.key.value },
            )
            .firstOrNull() ?: return null
        val completed = candidate.value.count { it.completedEpochDay != null }
        return WeeklyInsight.ReliableWeekday(
            "${weekdayVi(candidate.key)} ổn định nhất: hoàn thành $completed/${candidate.value.size} buổi đã xếp.",
        )
    }

    private fun difficultyInsight(
        evidence: List<WorkoutHistoryEntry>,
        feedback: List<WorkoutFeedback>,
    ): WeeklyInsight.DifficultyTrend? {
        val sessionIds = evidence.mapTo(mutableSetOf()) { it.sessionId }
        val relevant = feedback.filter { it.sessionId in sessionIds }
        if (relevant.size < 4) return null
        val hard = relevant.count { it.difficulty == WorkoutDifficulty.HARD }
        val easy = relevant.count { it.difficulty == WorkoutDifficulty.EASY }
        val message = when {
            hard * 2 >= relevant.size -> "$hard/${relevant.size} buổi được đánh giá quá nặng; nên ưu tiên hồi phục."
            easy * 2 >= relevant.size -> "$easy/${relevant.size} buổi được đánh giá quá nhẹ; chương trình có thể cần tăng dần."
            else -> "${relevant.size} phản hồi cho thấy độ khó nhìn chung vừa sức."
        }
        return WeeklyInsight.DifficultyTrend(message)
    }

    private fun timeBudgetInsight(
        evidence: List<WorkoutHistoryEntry>,
    ): WeeklyInsight.TimeBudgetPattern? {
        val shortened = evidence.filter { it.selectedTimeBudgetMinutes != null }
        if (shortened.size < 4) return null
        val common = shortened.groupingBy { it.selectedTimeBudgetMinutes!! }.eachCount()
            .entries.sortedWith(compareByDescending<Map.Entry<Int, Int>> { it.value }.thenBy { it.key }).first()
        return WeeklyInsight.TimeBudgetPattern(
            "Bạn đã chọn bản ${common.key} phút trong ${shortened.size} buổi gần đây.",
        )
    }

    private fun scheduleDriftInsight(
        evidence: List<WorkoutHistoryEntry>,
    ): WeeklyInsight.ScheduleDrift? {
        val completed = evidence.filter { it.completedEpochDay != null }
        if (completed.size < 3) return null
        val shifted = completed.filter { it.completedEpochDay != it.dueEpochDay }
        if (shifted.isEmpty()) return null
        return WeeklyInsight.ScheduleDrift(
            "${shifted.size}/${completed.size} buổi hoàn thành khác ngày dự kiến.",
        )
    }

    private fun weekStart(epochDay: Long): Long = LocalDate.ofEpochDay(epochDay)
        .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)).toEpochDay()

    private fun weekdayVi(day: DayOfWeek): String = when (day) {
        DayOfWeek.MONDAY -> "Thứ Hai"
        DayOfWeek.TUESDAY -> "Thứ Ba"
        DayOfWeek.WEDNESDAY -> "Thứ Tư"
        DayOfWeek.THURSDAY -> "Thứ Năm"
        DayOfWeek.FRIDAY -> "Thứ Sáu"
        DayOfWeek.SATURDAY -> "Thứ Bảy"
        DayOfWeek.SUNDAY -> "Chủ Nhật"
    }
}

File: flutter/lib/core/progress/weekly_insight_engine.dart
import '../model/feedback_models.dart';
import '../model/workout_models.dart';
import '../model/goal_models.dart';

sealed class WeeklyInsight {
  final String messageVi;
  WeeklyInsight(this.messageVi);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is WeeklyInsight &&
          runtimeType == other.runtimeType &&
          messageVi == other.messageVi;

  @override
  int get hashCode => messageVi.hashCode;
}

class WeeklyInsightAdherenceTrend extends WeeklyInsight {
  WeeklyInsightAdherenceTrend(super.messageVi);
}

class WeeklyInsightReliableWeekday extends WeeklyInsight {
  WeeklyInsightReliableWeekday(super.messageVi);
}

class WeeklyInsightDifficultyTrend extends WeeklyInsight {
  WeeklyInsightDifficultyTrend(super.messageVi);
}

class WeeklyInsightTimeBudgetPattern extends WeeklyInsight {
  WeeklyInsightTimeBudgetPattern(super.messageVi);
}

class WeeklyInsightScheduleDrift extends WeeklyInsight {
  WeeklyInsightScheduleDrift(super.messageVi);
}

class WeeklyInsightEngine {
  static List<WeeklyInsight> generate({
    required List<WorkoutHistoryEntry> history,
    required List<WorkoutFeedback> feedback,
    required int todayEpochDay,
  }) {
    final currentMonday = _mondayOfWeek(todayEpochDay);
    final windowStart = currentMonday - 28;

    final evidence = history
        .where((it) =>
            it.dueEpochDay >= windowStart && it.dueEpochDay < currentMonday)
        .toList();

    // Group by week start
    final byWeek = <int, List<WorkoutHistoryEntry>>{};
    for (var h in evidence) {
      final start = _mondayOfWeek(h.dueEpochDay);
      byWeek.putIfAbsent(start, () => []).add(h);
    }

    if (byWeek.keys.length < 2) return [];

    final sortedWeekStarts = byWeek.keys.toList()..sort();

    final insights = <WeeklyInsight>[];

    final adherence = _adherenceInsight(byWeek, sortedWeekStarts);
    if (adherence != null) insights.add(adherence);

    final reliable = _reliableWeekdayInsight(evidence);
    if (reliable != null) insights.add(reliable);

    final diff = _difficultyInsight(evidence, feedback);
    if (diff != null) insights.add(diff);

    final budget = _timeBudgetInsight(evidence);
    if (budget != null) insights.add(budget);

    final drift = _scheduleDriftInsight(evidence);
    if (drift != null) insights.add(drift);

    return insights.take(3).toList();
  }

  static WeeklyInsightAdherenceTrend? _adherenceInsight(
    Map<int, List<WorkoutHistoryEntry>> byWeek,
    List<int> sortedWeekStarts,
  ) {
    final rates = sortedWeekStarts.map((weekStart) {
      final week = byWeek[weekStart]!;
      final completedCount =
          week.where((it) => it.completedEpochDay != null).length;
      return completedCount * 100.0 / week.length;
    }).toList();

    final split = rates.length ~/ 2;
    if (split == 0) return null;

    final older =
        rates.take(split).fold<double>(0, (sum, val) => sum + val) / split;
    final newer = rates.skip(split).fold<double>(0, (sum, val) => sum + val) /
        (rates.length - split);

    final difference = (newer - older).round();
    if (difference.abs() < 15) return null;

    final direction = difference > 0 ? "tăng" : "giảm";
    return WeeklyInsightAdherenceTrend(
      "Tỷ lệ bám lịch $direction ${difference.abs()} điểm phần trăm qua ${rates.length} tuần đầy đủ.",
    );
  }

  static WeeklyInsightReliableWeekday? _reliableWeekdayInsight(
    List<WorkoutHistoryEntry> evidence,
  ) {
    final byWeekday = <WeekDay, List<WorkoutHistoryEntry>>{};
    for (var entry in evidence) {
      final date = DateTime.fromMillisecondsSinceEpoch(
        entry.dueEpochDay * 24 * 60 * 60 * 1000,
        isUtc: true,
      );
      final weekday = WeekDay.fromValue(date.weekday);
      byWeekday.putIfAbsent(weekday, () => []).add(entry);
    }

    final candidates =
        byWeekday.entries.where((e) => e.value.length >= 3).toList();
    if (candidates.isEmpty) return null;

    // Sắp xếp theo tỷ lệ hoàn thành giảm dần, sau đó theo weekday value tăng dần
    candidates.sort((a, b) {
      final aRate = a.value
              .where((it) => it.completedEpochDay != null)
              .length
              .toDouble() /
          a.value.length;
      final bRate = b.value
              .where((it) => it.completedEpochDay != null)
              .length
              .toDouble() /
          b.value.length;
      if (aRate != bRate) {
        return bRate.compareTo(aRate);
      }
      return a.key.value.compareTo(b.key.value);
    });

    final best = candidates.first;
    final completedCount =
        best.value.where((it) => it.completedEpochDay != null).length;
    return WeeklyInsightReliableWeekday(
      "${_weekdayVi(best.key)} ổn định nhất: hoàn thành $completedCount/${best.value.length} buổi đã xếp.",
    );
  }

  static WeeklyInsightDifficultyTrend? _difficultyInsight(
    List<WorkoutHistoryEntry> evidence,
    List<WorkoutFeedback> feedback,
  ) {
    final sessionIds = evidence.map((e) => e.sessionId).toSet();
    final relevant =
        feedback.where((f) => sessionIds.contains(f.sessionId)).toList();
    if (relevant.length < 4) return null;

    final hardCount =
        relevant.where((f) => f.difficulty == WorkoutDifficulty.hard).length;
    final easyCount =
        relevant.where((f) => f.difficulty == WorkoutDifficulty.easy).length;

    final String message;
    if (hardCount * 2 >= relevant.length) {
      message =
          "$hardCount/${relevant.length} buổi được đánh giá quá nặng; nên ưu tiên hồi phục.";
    } else if (easyCount * 2 >= relevant.length) {
      message =
          "$easyCount/${relevant.length} buổi được đánh giá quá nhẹ; chương trình có thể cần tăng dần.";
    } else {
      message =
          "${relevant.length} phản hồi cho thấy độ khó nhìn chung vừa sức.";
    }

    return WeeklyInsightDifficultyTrend(message);
  }

  static WeeklyInsightTimeBudgetPattern? _timeBudgetInsight(
    List<WorkoutHistoryEntry> evidence,
  ) {
    final shortened =
        evidence.where((e) => e.selectedTimeBudgetMinutes != null).toList();
    if (shortened.length < 4) return null;

    final counts = <int, int>{};
    for (var s in shortened) {
      final minutes = s.selectedTimeBudgetMinutes!;
      counts[minutes] = (counts[minutes] ?? 0) + 1;
    }

    final entries = counts.entries.toList();
    entries.sort((a, b) {
      if (a.value != b.value) {
        return b.value.compareTo(a.value); // Sắp xếp số lượng giảm dần
      }
      return a.key.compareTo(b.key); // Khớp phút tăng dần
    });

    final common = entries.first;
    return WeeklyInsightTimeBudgetPattern(
      "Bạn đã chọn bản ${common.key} phút trong ${shortened.length} buổi gần đây.",
    );
  }

  static WeeklyInsightScheduleDrift? _scheduleDriftInsight(
    List<WorkoutHistoryEntry> evidence,
  ) {
    final completed =
        evidence.where((e) => e.completedEpochDay != null).toList();
    if (completed.length < 3) return null;

    final shifted =
        completed.where((e) => e.completedEpochDay != e.dueEpochDay).toList();
    if (shifted.isEmpty) return null;

    return WeeklyInsightScheduleDrift(
      "${shifted.length}/${completed.length} buổi hoàn thành khác ngày dự kiến.",
    );
  }

  static int _mondayOfWeek(int epochDay) {
    final date = DateTime.fromMillisecondsSinceEpoch(
      epochDay * 24 * 60 * 60 * 1000,
      isUtc: true,
    );
    final monday = date.subtract(Duration(days: date.weekday - 1));
    return monday.millisecondsSinceEpoch ~/ (24 * 60 * 60 * 1000);
  }

  static String _weekdayVi(WeekDay day) {
    switch (day) {
      case WeekDay.monday:
        return "Thứ Hai";
      case WeekDay.tuesday:
        return "Thứ Ba";
      case WeekDay.wednesday:
        return "Thứ Tư";
      case WeekDay.thursday:
        return "Thứ Năm";
      case WeekDay.friday:
        return "Thứ Sáu";
      case WeekDay.saturday:
        return "Thứ Bảy";
      case WeekDay.sunday:
        return "Chủ Nhật";
    }
  }
}
